File: Day04/main.py
def read_lines():
    with open("input.txt") as file:
        input = file.read()
    return input.rstrip('\n').split('\n')

def transpose(lines):
    return [ ''.join(line) for line in zip(*lines) ]

def rotate(lines):
    return [ ''.join(line) for line in zip(*lines[::-1]) ]

def lower_diagonals(lines):
    n = len(lines)
    return [ ''.join([ lines[i][j] for (i, j) in zip(range(n - 1 - x, n), range(0, n)) ]) for x in range(0, n -1) ]

def upper_diagonals(lines):
    return lower_diagonals(transpose(lines))

def diagonals(lines):
    n = len(lines)
    return [ ''.join([lines[i][i] for i in range(n)]) ] + lower_diagonals(lines) + upper_diagonals(lines)

def symmetries(lines):
    line_symmetries = [ lines, transpose(lines), diagonals(lines), diagonals(rotate(lines)) ]
    return [ line for lines in line_symmetries for line in lines ] + [ line[::-1] for lines in line_symmetries for line in lines ]

def part1():
    lines = read_lines()
    answer = sum(line.count("XMAS") for line in symmetries(lines))
    print(answer)

def part2():
    lines = read_lines()
    n = len(lines)
    answer = 0
    for i in range(n - 2):
        for j in range(n - 2):
            if lines[i + 1][j + 1] != 'A':
                continue
            if { lines[i][j], lines[i + 2][j + 2] } != { 'M', 'S' }:
                continue
            if { lines[i + 2][j], lines[i][j + 2] } != { 'M', 'S' }:
                continue
            answer += 1
    print(answer)

if __name__ == '__main__':
    part1()
    part2()

File: Day04/main.hs
import Data.Array
import Prelude hiding (words)

type Grid = Array (Int, Int) Char

parseInput :: IO Grid
parseInput = do
  contents <- readFile "input.txt"
  let n = length (lines contents)
      bounds = ((1, 1), (n, n))
  return $ listArray bounds (concat (lines contents))

-- helper for all the words in direction (dx, dy) of length n
words :: Grid -> (Int, Int) -> Int -> [String]
words grid (dx, dy) n = [
  map (grid !) indices
    | (x, y) <- range (bounds grid)
    , let indices = [ (x + step * dx, y + step * dy) | step <- [0..n-1] ]
    , all (inRange (bounds grid)) indices
  ]

-- helper for all nxm windows
windows :: Grid -> (Int, Int) -> [Grid]
windows grid (n, m) = [
  listArray ((1, 1), (n, m)) (map (grid !) indices)
    | (x, y) <- range (bounds grid)
    , let indices = [ (x + dx, y + dy) | dx <- [0..n-1], dy <- [0..m-1] ]
    , all (inRange (bounds grid)) indices
  ]

part1 :: IO ()
part1 = do
  grid <- parseInput
  let directions = [ (dx, dy) | dx <- [-1, 0, 1], dy <- [-1, 0, 1], dx /= 0 || dy /= 0 ]
  let answer = length [ word | direction <- directions, word <- words grid direction 4, word == "XMAS" ]
  print answer

part2 :: IO ()
part2 = do
  grid <- parseInput
  let answer = length [
        window
          | window <- windows grid (3, 3)
          , "MAS" `elem` (words window (1,  1) 3 ++ words window (-1, -1) 3)
          , "MAS" `elem` (words window (1, -1) 3 ++ words window (-1,  1) 3)
        ]
  print answer

main :: IO ()
main = do
  part1
  part2
